src/pymor/algorithms/tr.py

```diff
@@ -0,0 +1,444 @@
+                    rejected = True
+                    # shrink the radius
+                    radius *= shrink_factor
+
+                    msg = 'Estimated output larger than previous output.'
+                else:
+                    surrogate.extend(mu)
+                    current_output = surrogate.new_rom_output(mu)
+                    if current_output <= compare_output:
+                        current_fom_output = surrogate.fom_output(mu)
+                        fom_output_diff = old_fom_output - current_fom_output
+                        rom_output_diff = old_rom_output - current_output
+                        if fom_output_diff >= radius_tol * rom_output_diff:
+                            # increase the radius if the model confidence is high enough
+                            radius /= shrink_factor
+
+                        msg = 'Updated model output smaller than previous output.'
+                    else:
+                        # reject new mu
+                        rejected = True
+                        # shrink the radius
+                        radius *= shrink_factor
+
+                        msg = 'Updated model output larger than previous output.'
+
+            # handle parameter rejection
+            if not rejected:
+                data['mus'].append(mu.copy())
+                mu_norm = np.linalg.norm(mu)
+                update_norms.append(np.linalg.norm(mu - old_mu))
+
+                data['subproblem_data'].append(sub_data)
+
+                with logger.block('Computing first order criticality...'):
+                    gradient = surrogate.fom_gradient(mu)
+                    first_order_criticality = np.linalg.norm(mu - parameter_space.clip(mu - gradient).to_numpy())
+                    foc_norms.append(first_order_criticality)
+
+                surrogate.accept()
+                logger.info(f'Current mu iterate accepted: {msg}')
+
+                old_rom_output = current_output
+            else:
+                mu = old_mu
+                surrogate.reject()
+                logger.info(f'Current mu iterate rejected: {msg}')
+
+            with warnings.catch_warnings():
+                # ignore division-by-zero warnings when solution_norm or output is zero
+                warnings.filterwarnings('ignore', category=RuntimeWarning)
+                logger.info(f'it:{iteration} '
+                            f'foc:{first_order_criticality:.3e} '
+                            f'radius:{radius:.3e}')
+
+            if not np.isfinite(mu_norm):
+                raise TRError('Failed to converge.')
+
+    logger.info('')
+
+    data['update_norms'] = np.array(update_norms)
+    data['foc_norms'] = np.array(foc_norms)
+    data['iterations'] = iteration
+    data['fom_evaluations'] = surrogate.fom_evaluations
+    data['rom_evaluations'] = surrogate.rom_evaluations
+    data['enrichments'] = surrogate.enrichments
+
+    return mu, data
+
+
+class TRSurrogate(BasicObject):
```

Comment:
Somehow the name `Surrogate` does not seem quite fitting. I would not expect a surrogate expect to have methods like `fom_output` or `fom_gradient`. Is there another option for a name?
Actually, I am wondering whether it would make more sense to actually remove these methods and have the fom as a parameter of the tr algorithms. Would make sense to me that I need to give the algorithm a fom and a surrogate.

Reply:
I agree with the latter. Probably the `surrogate` should not have fom methods.

Reply:
Maybe I misunderstand what you are trying to achieve here. But, what I have in mind is to remove the fom from the surrogate altogether. (Of course, depending on the impl of the surrogate, there might be a fom in the background, but that should not be part of the interface.) Instead, the fom would be passed as an additional argument to `trust_region`. The fom evaluations could be counted where they occur.

Reply:
Oh right, I now pass the FOM directly to the `trust_region` method as well as a method argument if we need it within the surrogate (except for the `__init__` where I use the reductor as a work around instead.